File: src/iCAX-UI/SDK/Resources/resourceURL.mjs
const canonicalGuidPattern =
  /^[0-9a-f]{8}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{4}-[0-9a-f]{12}$/;
const safeIdPattern = /^[A-Za-z0-9._-]+$/;

export function generateResourceId() {
  if (typeof globalThis.crypto?.randomUUID === "function") {
    return globalThis.crypto.randomUUID().toLowerCase();
  }

  const bytes = new Uint8Array(16);
  if (typeof globalThis.crypto?.getRandomValues === "function") {
    globalThis.crypto.getRandomValues(bytes);
  } else {
    for (let index = 0; index < bytes.length; index += 1) {
      bytes[index] = Math.floor(Math.random() * 256);
    }
  }
  bytes[6] = (bytes[6] & 0x0f) | 0x40;
  bytes[8] = (bytes[8] & 0x3f) | 0x80;
  const text = [...bytes]
    .map((value) => value.toString(16).padStart(2, "0"))
    .join("");
  return [
    text.slice(0, 8),
    text.slice(8, 12),
    text.slice(12, 16),
    text.slice(16, 20),
    text.slice(20),
  ].join("-");
}

export function makeResourceCollectionURL(scope = {}) {
  const applicationId = requireSafeId(
    scope.applicationId,
    "applicationId",
  );
  let url = `resource://${applicationId}`;

  if (scope.productId !== undefined && scope.productId !== null && scope.productId !== "") {
    url += `/products/${requireSafeId(scope.productId, "productId")}`;
  } else {
    requireAbsent(scope.projectId, "projectId");
    requireAbsent(scope.sceneId, "sceneId");
    return `${url}/resources`;
  }

  if (scope.projectId) {
    url += `/projects/${requireGuid(scope.projectId, "projectId")}`;
  } else {
    requireAbsent(scope.sceneId, "sceneId");
    return `${url}/resources`;
  }

  if (scope.sceneId) {
    url += `/scenes/${requireGuid(scope.sceneId, "sceneId")}`;
  }
  return `${url}/resources`;
}

export function makeResourceURL(scope, resourceId) {
  return `${makeResourceCollectionURL(scope)}/${requireGuid(resourceId, "resourceId")}`;
}

export function allocateResourceURL(scope) {
  const resourceId = generateResourceId();
  return {
    resourceId,
    url: makeResourceURL(scope, resourceId),
  };
}

export function parseResourceURL(value) {
  const text = String(value ?? "");
  const prefix = "resource://";
  if (!text.startsWith(prefix) || text.includes("?") || text.includes("#")) {
    throw new TypeError("resource URL is not canonical");
  }

  const remainder = text.slice(prefix.length);
  const slash = remainder.indexOf("/");
  if (slash <= 0) {
    throw new TypeError("resource URL has no resources collection");
  }
  const path = remainder.slice(slash + 1);
  if (!path || path.endsWith("/")) {
    throw new TypeError("resource URL contains an empty path segment");
  }
  const segments = path.split("/");
  if (segments.some((segment) => !segment)) {
    throw new TypeError("resource URL contains an empty path segment");
  }

  const result = {
    applicationId: requireSafeId(remainder.slice(0, slash), "applicationId"),
    productId: "",
    projectId: "",
    sceneId: "",
    resourceId: "",
  };
  let index = 0;
  if (segments[index] === "products") {
    result.productId = requireSafeId(segments[index + 1], "productId");
    index += 2;
  }
  if (segments[index] === "projects") {
    if (!result.productId) {
      throw new TypeError("project resource URL requires productId");
    }
    result.projectId = requireGuid(segments[index + 1], "projectId");
    index += 2;
  }
  if (segments[index] === "scenes") {
    if (!result.projectId) {
      throw new TypeError("scene resource URL requires projectId");
    }
    result.sceneId = requireGuid(segments[index + 1], "sceneId");
    index += 2;
  }
  if (segments[index] !== "resources") {
    throw new TypeError("resource URL has no resources collection");
  }
  index += 1;
  if (index < segments.length) {
    result.resourceId = requireGuid(segments[index], "resourceId");
    index += 1;
  }
  if (index !== segments.length) {
    throw new TypeError("resource URL has unexpected trailing segments");
  }
  result.scope = result.sceneId
    ? "scene"
    : result.projectId
      ? "project"
      : result.productId
        ? "product"
        : "application";
  result.isCollection = !result.resourceId;
  return result;
}

function requireSafeId(value, name) {
  const text = String(value ?? "");
  if (!safeIdPattern.test(text) || text === "." || text === ".." || text.includes("..")) {
    throw new TypeError(`${name} is not a URL-safe stable ID`);
  }
  return text;
}

function requireGuid(value, name) {
  const text = String(value ?? "");
  if (!canonicalGuidPattern.test(text)
    || text === "00000000-0000-0000-0000-000000000000") {
    throw new TypeError(`${name} must be a canonical non-nil GUID`);
  }
  return text;
}

function requireAbsent(value, name) {
  if (value !== undefined && value !== null && value !== "") {
    throw new TypeError(`${name} requires its parent resource scope`);
  }
}

File: src/iCAX-Engine/framework/Resources/ResourceURL.cpp
#include "pch.h"
#include "ResourceURL.h"

#include <stdexcept>
#include <vector>

namespace
{
    using namespace iCAX::Resource;

    bool IsSafeIDSegment(IN const std::string& strValue_) noexcept
    {
        if (strValue_.empty() ||
            strValue_.find("..") != std::string::npos)
        {
            return false;
        }
        for (const auto _Character : strValue_)
        {
            const bool _bDigit =
                _Character >= '0' && _Character <= '9';
            const bool _bLower =
                _Character >= 'a' && _Character <= 'z';
            const bool _bUpper =
                _Character >= 'A' && _Character <= 'Z';
            const bool _bSymbol =
                _Character == '.' ||
                _Character == '_' ||
                _Character == '-';
            if (!_bDigit && !_bLower && !_bUpper && !_bSymbol)
            {
                return false;
            }
        }
        return true;
    }

    std::vector<std::string> SplitPath(
        IN const std::string& strPath_)
    {
        if (strPath_.empty() || strPath_.ends_with('/'))
        {
            throw std::invalid_argument(
                "Resource URL contains an empty path segment");
        }
        std::vector<std::string> _Segments;
        std::size_t _Begin = 0;
        while (_Begin < strPath_.size())
        {
            const auto _End =
                strPath_.find('/', _Begin);
            auto _Segment = strPath_.substr(
                _Begin,
                _End == std::string::npos
                    ? std::string::npos
                    : _End - _Begin);
            if (_Segment.empty())
            {
                throw std::invalid_argument(
                    "Resource URL contains an empty path segment");
            }
            _Segments.push_back(std::move(_Segment));
            if (_End == std::string::npos)
            {
                break;
            }
            _Begin = _End + 1;
        }
        return _Segments;
    }

    iCAX::Data::uuid ParseCanonicalUUID(
        IN const std::string& strValue_,
        IN const char* pName_)
    {
        const auto _ID =
            iCAX::Data::uuid::from_string(strValue_);
        if (!_ID ||
            _ID->is_nil() ||
            iCAX::Data::to_string(*_ID) != strValue_)
        {
            throw std::invalid_argument(
                std::string(pName_) +
                " must be a canonical non-nil GUID");
        }
        return *_ID;
    }

    void RequireScope(IN const CResourceScope& Scope_)
    {
        if (!Scope_.IsValid())
        {
            throw std::invalid_argument(
                "Resource scope is invalid");
        }
    }
}

bool iCAX::Resource::CResourceScope::IsValid() const noexcept
{
    if (!IsSafeIDSegment(ApplicationID))
    {
        return false;
    }

    switch (Scope)
    {
    case EResourceScope::Application:
        return ProductID.empty() &&
            ProjectID.is_nil() &&
            SceneID.is_nil();
    case EResourceScope::Product:
        return IsSafeIDSegment(ProductID) &&
            ProjectID.is_nil() &&
            SceneID.is_nil();
    case EResourceScope::Project:
        return IsSafeIDSegment(ProductID) &&
            !ProjectID.is_nil() &&
            SceneID.is_nil();
    case EResourceScope::Scene:
        return IsSafeIDSegment(ProductID) &&
            !ProjectID.is_nil() &&
            !SceneID.is_nil();
    default:
        return false;
    }
}

bool iCAX::Resource::CResourceURL::IsCollection() const noexcept
{
    return Owner.IsValid() && ResourceID.is_nil();
}

bool iCAX::Resource::CResourceURL::IsResource() const noexcept
{
    return Owner.IsValid() && !ResourceID.is_nil();
}

iCAX::Resource::CResourceScope
iCAX::Resource::MakeApplicationResourceScope(
    IN const std::string& strApplicationID_)
{
    CResourceScope _Scope;
    _Scope.Scope = EResourceScope::Application;
    _Scope.ApplicationID = strApplicationID_;
    RequireScope(_Scope);
    return _Scope;
}

iCAX::Resource::CResourceScope
iCAX::Resource::MakeProductResourceScope(
    IN const std::string& strApplicationID_,
    IN const std::string& strProductID_)
{
    CResourceScope _Scope;
    _Scope.Scope = EResourceScope::Product;
    _Scope.ApplicationID = strApplicationID_;
    _Scope.ProductID = strProductID_;
    RequireScope(_Scope);
    return _Scope;
}

iCAX::Resource::CResourceScope
iCAX::Resource::MakeProjectResourceScope(
    IN const std::string& strApplicationID_,
    IN const std::string& strProductID_,
    IN const iCAX::Data::uuid& ProjectID_)
{
    CResourceScope _Scope;
    _Scope.Scope = EResourceScope::Project;
    _Scope.ApplicationID = strApplicationID_;
    _Scope.ProductID = strProductID_;
    _Scope.ProjectID = ProjectID_;
    RequireScope(_Scope);
    return _Scope;
}

iCAX::Resource::CResourceScope
iCAX::Resource::MakeSceneResourceScope(
    IN const std::string& strApplicationID_,
    IN const std::string& strProductID_,
    IN const iCAX::Data::uuid& ProjectID_,
    IN const iCAX::Data::uuid& SceneID_)
{
    CResourceScope _Scope;
    _Scope.Scope = EResourceScope::Scene;
    _Scope.ApplicationID = strApplicationID_;
    _Scope.ProductID = strProductID_;
    _Scope.ProjectID = ProjectID_;
    _Scope.SceneID = SceneID_;
    RequireScope(_Scope);
    return _Scope;
}

iCAX::Data::uuid iCAX::Resource::GenerateResourceID()
{
    return iCAX::Data::GenerateNewUUID();
}

std::string iCAX::Resource::MakeResourceCollectionURL(
    IN const CResourceScope& Scope_)
{
    RequireScope(Scope_);

    auto _URL =
        std::string("resource://") +
        Scope_.ApplicationID;
    if (Scope_.Scope >= EResourceScope::Product)
    {
        _URL += "/products/" + Scope_.ProductID;
    }
    if (Scope_.Scope >= EResourceScope::Project)
    {
        _URL +=
            "/projects/" +
            iCAX::Data::to_string(Scope_.ProjectID);
    }
    if (Scope_.Scope >= EResourceScope::Scene)
    {
        _URL +=
            "/scenes/" +
            iCAX::Data::to_string(Scope_.SceneID);
    }
    return _URL + "/resources";
}

std::string iCAX::Resource::MakeResourceURL(
    IN const CResourceScope& Scope_,
    IN const iCAX::Data::uuid& ResourceID_)
{
    if (ResourceID_.is_nil())
    {
        throw std::invalid_argument(
            "ResourceID cannot be nil");
    }
    return MakeResourceCollectionURL(Scope_) +
        "/" +
        iCAX::Data::to_string(ResourceID_);
}

iCAX::Resource::CResourceURL
iCAX::Resource::ParseResourceURL(
    IN const std::string& strURL_)
{
    constexpr const char* kPrefix = "resource://";
    if (!strURL_.starts_with(kPrefix))
    {
        throw std::invalid_argument(
            "Resource URL must use resource://");
    }
    if (strURL_.find_first_of("?#") != std::string::npos)
    {
        throw std::invalid_argument(
            "Resource URL cannot contain query or fragment");
    }

    const auto _Remainder =
        strURL_.substr(std::char_traits<char>::length(kPrefix));
    const auto _Slash = _Remainder.find('/');
    if (_Slash == std::string::npos)
    {
        throw std::invalid_argument(
            "Resource URL has no resources path");
    }

    const auto _ApplicationID =
        _Remainder.substr(0, _Slash);
    if (!IsSafeIDSegment(_ApplicationID))
    {
        throw std::invalid_argument(
            "Resource URL ApplicationID is invalid");
    }
    const auto _Segments =
        SplitPath(_Remainder.substr(_Slash + 1));

    CResourceURL _Result;
    std::size_t _ResourceIndex = 0;
    if (_Segments.size() >= 1 &&
        _Segments[0] == "resources")
    {
        _Result.Owner =
            MakeApplicationResourceScope(_ApplicationID);
        _ResourceIndex = 1;
    }
    else if (_Segments.size() >= 3 &&
        _Segments[0] == "products" &&
        _Segments[2] == "resources")
    {
        _Result.Owner = MakeProductResourceScope(
            _ApplicationID,
            _Segments[1]);
        _ResourceIndex = 3;
    }
    else if (_Segments.size() >= 5 &&
        _Segments[0] == "products" &&
        _Segments[2] == "projects" &&
        _Segments[4] == "resources")
    {
        _Result.Owner = MakeProjectResourceScope(
            _ApplicationID,
            _Segments[1],
            ParseCanonicalUUID(
                _Segments[3],
                "ProjectID"));
        _ResourceIndex = 5;
    }
    else if (_Segments.size() >= 7 &&
        _Segments[0] == "products" &&
        _Segments[2] == "projects" &&
        _Segments[4] == "scenes" &&
        _Segments[6] == "resources")
    {
        _Result.Owner = MakeSceneResourceScope(
            _ApplicationID,
            _Segments[1],
            ParseCanonicalUUID(
                _Segments[3],
                "ProjectID"),
            ParseCanonicalUUID(
                _Segments[5],
                "SceneID"));
        _ResourceIndex = 7;
    }
    else
    {
        throw std::invalid_argument(
            "Resource URL hierarchy is invalid");
    }

    if (_Segments.size() == _ResourceIndex)
    {
        return _Result;
    }
    if (_Segments.size() != _ResourceIndex + 1)
    {
        throw std::invalid_argument(
            "Resource URL has unexpected trailing segments");
    }

    _Result.ResourceID = ParseCanonicalUUID(
        _Segments[_ResourceIndex],
        "ResourceID");
    return _Result;
}

std::optional<iCAX::Resource::CResourceURL>
iCAX::Resource::TryParseResourceURL(
    IN const std::string& strURL_) noexcept
{
    try
    {
        return ParseResourceURL(strURL_);
    }
    catch (...)
    {
        return std::nullopt;
    }
}

bool iCAX::Resource::ResourceScopeEquals(
    IN const CResourceScope& Left_,
    IN const CResourceScope& Right_) noexcept
{
    return Left_.Scope == Right_.Scope &&
        Left_.ApplicationID == Right_.ApplicationID &&
        Left_.ProductID == Right_.ProductID &&
        Left_.ProjectID == Right_.ProjectID &&
        Left_.SceneID == Right_.SceneID;
}
